### PR TITLE
TEST-55-OOMD: use 'stress' to create high pressure, ensure swap is set up correctly on Ubuntu CI

### PR DESCRIPTION
**Pull request by bluca**
The test fails on my machine, running Debian stable, because testsuite-55-testbloat.service just swaps and never goes over the limit, so it's not killed.

### CODE REVIEW

**Review by yuwata — APPROVED** [review]

**Comment by bluca** [comment]
> The test fails on my machine, running Debian stable, because testsuite-55-testbloat.service just swaps and never goes over the limit, so it's not killed. Disable swap on the slice.
> 
> Maybe I am misunderstanding how this is supposed to work, or there's some missing thing on my host? But it doesn't work in qemu nor nspawn.

Journal of test run on debian stable, just the normal `sudo make -C test/TEST-55-OOMD/`

[TEST-55-OOMD.journal.zip](https://github.com/systemd/systemd/files/11779838/TEST-55-OOMD.journal.zip)

**Comment by poettering** [comment]
this smells weird, oomd should be *better* at doing its thing when swap is available, not worse...

@bluca what kernel is this?

**Comment by bluca** [comment]
> this smells weird, oomd should be _better_ at doing its thing when swap is available, not worse...
> 
> @bluca what kernel is this?

I know it's weird, but it is not even flaky, it just doesn't work without this change. It's debian stable so 6.1.27

**Comment by anitazha** [comment]
I think your suspicion is right that the testbloat service is swapping more on Debian, probably as a result of using the system using swap partition instead of swap on zram like Fedora. So the testbloat service remains under its memory.high value and thus will not throttle. Disabling swap at the test cgroup level works to prevent this and thus brings back the throttling which creates pressure.

Another way to fix this is to write testsuite-55-slowgrowth.sh better such that it actually uses the memory it allocates so that it doesn't get easily swapped out. I could try taking a stab at that. But otherwise there's nothing inherently wrong with merging this.

**Comment by mrc0mmand** [comment]
> It seems to work more reliably by switching to 'stress', using `stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1` I get high memory pressure even with swap. @mrc0mmand could you please add that to centos?

Sure, done. But you'll have to force-push to restart the CIs since I can't do that for some reason.

**Comment by bluca** [comment]
aaand of course this change makes it worse in the CI -_-

@anitazha it is strange as it is clearly above the limit:

```
8225s [ 2158.602300] testsuite-55.sh[59]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
8225s [ 2158.635937] testsuite-55.sh[746]: ● testsuite-55-testbloat.service - Create a lot of memory pressure
8225s [ 2158.635987] testsuite-55.sh[746]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
8225s [ 2158.636050] testsuite-55.sh[746]:      Active: active (running) since Wed 2023-06-21 04:00:31 UTC; 2min 0s ago
8225s [ 2158.636071] testsuite-55.sh[746]:    Main PID: 173
8225s [ 2158.636087] testsuite-55.sh[746]:       Tasks: 11 (limit: 4606)
8225s [ 2158.636103] testsuite-55.sh[746]:      Memory: 8.0M (high: 5.0M available: 0B)
8225s [ 2158.636126] testsuite-55.sh[746]:         CPU: 173ms
8225s [ 2158.636143] testsuite-55.sh[746]:      CGroup: /user.slice/user-4711.slice/user@4711.service/testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testbloat.service
8225s [ 2158.636165] testsuite-55.sh[746]:              ├─173 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636182] testsuite-55.sh[746]:              ├─174 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636202] testsuite-55.sh[746]:              ├─175 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636218] testsuite-55.sh[746]:              ├─176 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636239] testsuite-55.sh[746]:              ├─177 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636261] testsuite-55.sh[746]:              ├─178 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636278] testsuite-55.sh[746]:              ├─179 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636299] testsuite-55.sh[746]:              ├─180 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636315] testsuite-55.sh[746]:              ├─181 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636338] testsuite-55.sh[746]:              ├─182 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.636358] testsuite-55.sh[746]:              └─183 /usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
8225s [ 2158.637655] testsuite-55.sh[59]: + exit 42
```

Any idea?

**Comment by anitazha** [comment]
Gonna give it a try on my laptop. It's killing properly on the system slices but failing on the user manager slice. My best guess is it needs more cool down before it attempts to run the next test.

File: test/units/testsuite-55-testbloat.service [hunk 1]
```diff
@@ -3,8 +3,8 @@
 Description=Create a lot of memory pressure
 
 [Service]
-# A VERY small memory.high will cause the script (trying to use a lot of memory)
+# A VERY small memory.high will cause the 'stress' (trying to use a lot of memory)
 # to throttle and be put under heavy pressure.
 MemoryHigh=5M
 Slice=testsuite-55-workload.slice
-ExecStart=/usr/lib/systemd/tests/testdata/units/testsuite-55-slowgrowth.sh
+ExecStart=/usr/bin/stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
```
Comment (keszybz) [inline test/units/testsuite-55-testbloat.service]: `ExecStart=stress …`

**Comment by bluca** [comment]
> Gonna give it a try on my laptop. It's killing properly on the system slices but failing on the user manager slice. My best guess is it needs more cool down before it attempts to run the next test.

I've added a sleep between tests, let's see if that helps

**Comment by bluca** [comment]
> > Gonna give it a try on my laptop. It's killing properly on the system slices but failing on the user manager slice. My best guess is it needs more cool down before it attempts to run the next test.
> 
> I've added a sleep between tests, let's see if that helps

still failing, any other ideas?

**Comment by bluca** [comment]
@anitazha there is something definitely going wrong somewhere for the user slice:

```
5002s [ 2333.123126] testsuite-55.sh[58]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
5003s [ 2333.162245] testsuite-55.sh[796]: ● testsuite-55-testbloat.service - Create a lot of memory pressure
5003s [ 2333.162293] testsuite-55.sh[796]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
5003s [ 2333.162311] testsuite-55.sh[796]:      Active: active (running) since Fri 2023-06-23 12:54:23 UTC; 1min 58s ago
5003s [ 2333.162335] testsuite-55.sh[796]:    Main PID: 174
5003s [ 2333.162352] testsuite-55.sh[796]:       Tasks: 11 (limit: 4606)
5003s [ 2333.162369] testsuite-55.sh[796]:      Memory: 5.6M (high: 3.0M available: 0B)
5003s [ 2333.162387] testsuite-55.sh[796]:         CPU: 179ms
5003s [ 2333.162405] testsuite-55.sh[796]:      CGroup: /user.slice/user-4711.slice/user@4711.service/testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testbloat.service
5003s [ 2333.162423] testsuite-55.sh[796]:              ├─174 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162440] testsuite-55.sh[796]:              ├─175 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162458] testsuite-55.sh[796]:              ├─176 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162474] testsuite-55.sh[796]:              ├─177 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162491] testsuite-55.sh[796]:              ├─178 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162508] testsuite-55.sh[796]:              ├─179 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162526] testsuite-55.sh[796]:              ├─180 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162550] testsuite-55.sh[796]:              ├─181 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162568] testsuite-55.sh[796]:              ├─182 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162584] testsuite-55.sh[796]:              ├─183 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162601] testsuite-55.sh[796]:              └─184 stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1
5003s [ 2333.162993] testsuite-55.sh[58]: + oomctl
5003s [ 2333.170239] testsuite-55.sh[800]: Dry Run: no
5003s [ 2333.170303] testsuite-55.sh[800]: Swap Used Limit: 90.00%
5003s [ 2333.170337] testsuite-55.sh[800]: Default Memory Pressure Limit: 60.00%
5003s [ 2333.170364] testsuite-55.sh[800]: Default Memory Pressure Duration: 2s
5003s [ 2333.170397] testsuite-55.sh[800]: System Context:
5003s [ 2333.170427] testsuite-55.sh[800]:      Memory: Used: 490.9M Total: 3.8G
5003s [ 2333.170468] testsuite-55.sh[800]:      Swap: Used: 0B Total: 0B
5003s [ 2333.170505] testsuite-55.sh[800]: Swap Monitored CGroups:
5003s [ 2333.170535] testsuite-55.sh[800]: Memory Pressure Monitored CGroups:
5003s [ 2333.170564] testsuite-55.sh[800]:      Path: /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice
5003s [ 2333.170603] testsuite-55.sh[800]:              Memory Pressure Limit: 20.00%
5003s [ 2333.170641] testsuite-55.sh[800]:              Pressure: Avg10: 0.00 Avg60: 0.19 Avg300: 0.99 Total: 6s
5003s [ 2333.170681] testsuite-55.sh[800]:              Current Memory Usage: 284.0K
5003s [ 2333.170721] testsuite-55.sh[800]:              Memory Min: 0B
5003s [ 2333.170760] testsuite-55.sh[800]:              Memory Low: 0B
5003s [ 2333.170798] testsuite-55.sh[800]:              Pgscan: 187
5003s [ 2333.170838] testsuite-55.sh[800]:              Last Pgscan: 187
5003s [ 2333.170878] testsuite-55.sh[800]:      Path: /user.slice/user-4711.slice/user@4711.service/testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice
5003s [ 2333.170914] testsuite-55.sh[800]:              Memory Pressure Limit: 20.00%
5003s [ 2333.170952] testsuite-55.sh[800]:              Pressure: Avg10: 99.00 Avg60: 85.08 Avg300: 32.94 Total: 1min 58s
5003s [ 2333.170991] testsuite-55.sh[800]:              Current Memory Usage: 5.8M
5003s [ 2333.171031] testsuite-55.sh[800]:              Memory Min: 0B
5003s [ 2333.171070] testsuite-55.sh[800]:              Memory Low: 0B
5003s [ 2333.171109] testsuite-55.sh[800]:              Pgscan: 0
5003s [ 2333.171150] testsuite-55.sh[800]:              Last Pgscan: 0
```

99% pressure for 2 minutes, but oomd doesn't kick in - only debug logs from oomd are about incoming messages on varlink. 
But it's so weird that it works on the CI without this changes, and adding _more_ pressure makes it _not_ work?

**Comment by bluca** [comment]
Same symptom in this PR for the system slice https://github.com/systemd/systemd/pull/27890 when running under sanitizers - so maybe it's a timing issue of some sort? When things are slower this happens?

```
15:44:01 [ 3336.080410] testsuite-55.sh[79]: + oomctl
15:44:01 [ 3336.100446] testsuite-55.sh[534]: Dry Run: no
15:44:01 [ 3336.100534] testsuite-55.sh[534]: Swap Used Limit: 90.00%
15:44:01 [ 3336.100591] testsuite-55.sh[534]: Default Memory Pressure Limit: 60.00%
15:44:01 [ 3336.100655] testsuite-55.sh[534]: Default Memory Pressure Duration: 2s
15:44:01 [ 3336.100711] testsuite-55.sh[534]: System Context:
15:44:01 [ 3336.100773] testsuite-55.sh[534]: 	Memory: Used: 2.8G Total: 125.7G
15:44:01 [ 3336.100831] testsuite-55.sh[534]: 	Swap: Used: 0B Total: 0B
15:44:01 [ 3336.100894] testsuite-55.sh[534]: Swap Monitored CGroups:
15:44:01 [ 3336.100947] testsuite-55.sh[534]: Memory Pressure Monitored CGroups:
15:44:01 [ 3336.101001] testsuite-55.sh[534]: 	Path: /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice
15:44:01 [ 3336.101070] testsuite-55.sh[534]: 		Memory Pressure Limit: 20.00%
15:44:01 [ 3336.101126] testsuite-55.sh[534]: 		Pressure: Avg10: 96.42 Avg60: 80.23 Avg300: 31.10 Total: 1min 55s
15:44:01 [ 3336.101187] testsuite-55.sh[534]: 		Current Memory Usage: 13.6M
15:44:01 [ 3336.101247] testsuite-55.sh[534]: 		Memory Min: 0B
15:44:01 [ 3336.101300] testsuite-55.sh[534]: 		Memory Low: 0B
15:44:01 [ 3336.101355] testsuite-55.sh[534]: 		Pgscan: 0
15:44:01 [ 3336.101409] testsuite-55.sh[534]: 		Last Pgscan: 0
15:44:01 [ 3336.107824] testsuite-55.sh[79]: + sleep 2
15:44:01 [ 3338.111893] testsuite-55.sh[537]: ++ date -u +%s
15:44:01 [ 3338.114227] testsuite-55.sh[79]: + [[ 1687531347 -le 1687531346 ]]
15:44:01 [ 3338.114506] testsuite-55.sh[79]: + systemctl status testsuite-55-testbloat.service
15:44:01 [ 3338.159282] testsuite-55.sh[538]: ● testsuite-55-testbloat.service - Create a lot of memory pressure
15:44:01 [ 3338.159484] testsuite-55.sh[538]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
15:44:01 [ 3338.159556] testsuite-55.sh[538]:      Active: active (running) since Fri 2023-06-23 14:40:26 UTC; 2min 0s ago
15:44:01 [ 3338.159625] testsuite-55.sh[538]:    Main PID: 176 (9)
15:44:01 [ 3338.159695] testsuite-55.sh[538]:       Tasks: 1 (limit: 154542)
15:44:01 [ 3338.159752] testsuite-55.sh[538]:      Memory: 7.4M (high: 3.0M available: 0B)
15:44:01 [ 3338.159822] testsuite-55.sh[538]:         CPU: 36ms
15:44:01 [ 3338.159889] testsuite-55.sh[538]:      CGroup: /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testbloat.service
15:44:01 [ 3338.159949] testsuite-55.sh[538]:              └─176 /usr/lib/systemd/systemd-executor --deserialize 38 --log-level debug --log-target journal
```

https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/3492/console

**Comment by bluca** [comment]
So the problem is that in both cases there are no reclaims - pgscan is 0 - so oomd does not react

**Comment by bluca** [comment]
Looks like the main issue was that on ubuntu swap was not set up correctly, so there are 0 pgscans

**Comment by bluca** [comment]
@anitazha this looks good for me now, problems in both CI and local runs are all solved

**Review by anitazha — APPROVED** [review]
LGTM

File: test/TEST-55-OOMD/test.sh [hunk 2]
```diff
@@ -5,18 +5,19 @@ set -e
...
+    # Create a swap file
     (
-        mkswap "${LOOPDEV:?}p2"
-        image_install swapon swapoff
+        image_install mkswap swapon swapoff stress
 
-        cat >>"${initdir:?}/etc/fstab" <<EOF
-UUID=$(blkid -o value -s UUID "${LOOPDEV}p2")    none    swap    defaults 0 0
-EOF
+        dd if=/dev/zero of="${initdir:?}/swapfile" bs=1M count=48
```
Comment (poettering) [inline test/TEST-55-OOMD/test.sh:19]: why not just call `truncate`?

Comment (bluca) [reply test/TEST-55-OOMD/test.sh:19]: just following the manpage suggestion

Comment (mrc0mmand) [reply test/TEST-55-OOMD/test.sh:19]: Also, `mkswap` will complain if you use a sparse file:

```
$ truncate -s 50M ~/tmp/foo
$ mkswap ~/tmp/foo 
mkswap: /home/fsumsal/tmp/foo: insecure permissions 0644, fix with: chmod 0600 /home/fsumsal/tmp/foo

mkswap: /home/fsumsal/tmp/foo contains holes or other unsupported extents.
        This swap file can be rejected by kernel on swap activation!
        Use --verbose for more details.

Setting up swapspace version 1, size = 50 MiB (52424704 bytes)
no label, UUID=e7e51d40-a2d5-4215-a56b-2bee4c1da7e9

```

**Pull request merged**